### PR TITLE
Feature/repositories

### PR DESCRIPTION
**Pull request by mattgen88**
By pulling out the queries into a series of services under infrastructure, we can restructure to use interfaces and easily test our database queries, or mock the infrastructure repositories in other code.

### CODE REVIEW

**Comment by Sigafoos** [comment]
We need to upgrade to 1.14 for this

File: exlibris.postman_collection.json [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+	},
+	"item": [
+		{
+			"name": "register",
+			"request": {
+				"method": "POST",
+				"header": [],
+				"body": {
+					"mode": "raw",
+					"raw": "{\n\t\"username\": \"mattgen88\",\n\t\"password\": \"password\",\n\t\"email\": \"matthew.general@gmail.com\",\n\t\"display_name\": \"matt\"\n}",
```
Comment (Sigafoos) [inline exlibris.postman_collection.json:15]: This should use postman variables

File: handler/handler.go [hunk 2]
```diff
@@ -3,29 +3,51 @@ package handler
...
-	cfg             *config.Config
-	bookService     *service.Book
-	authorService   *service.Author
-	editionsService *service.Editions
+	ap                   *activitypub.ActivityPub
+	cfg                  *config.Config
+	bookService          *service.Book
+	authorService        *service.Author
+	editionsService      *service.Editions
+	booksRepo            *books.Repository
```
Comment (Sigafoos) [inline handler/handler.go:26]: Do we need the `Repo` in each of these?

File: handler/auth.go [hunk 3]
```diff
@@ -141,8 +130,11 @@ func (h *Handler) Authenticate(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	var user model.User
-	h.db.First(&user, "username = ?", request.Username)
+	user, err := h.usersRepo.GetByUsername(request.Username)
+	if err != nil {
+		// Something went wrong
+		w.WriteHeader(http.StatusInternalServerError)
```
Comment (Sigafoos) [inline handler/auth.go:136]: Will this 500 on a not found?

File: handler/auth.go [hunk 4]
```diff
@@ -206,16 +196,7 @@ func (h *Handler) VerifyKey(w http.ResponseWriter, r *http.Request) {
...
-		registrationKey.User.Verified = true
-		if err := tx.Save(&registrationKey.User).Error; err != nil {
-			return err
-		}
-		if err := tx.Delete(&registrationKey).Error; err != nil {
-			return err
-		}
-		return nil
-	}); err != nil {
+	if err := h.usersRepo.Activate(&registrationKey.User, registrationKey); err != nil {
```
Comment (Sigafoos) [inline handler/auth.go:199]: I think I’d like this better as 

```go
h.usersRepo.Activate(key)
```

And let it handle loading the user, etc

**Comment by Sigafoos** [comment]
Incorporate #122 once it's merged

File: handler/auth.go (hunk 3 above)
Comment (mattgen88) [reply handler/auth.go:136]: yes, it would. I'll check for users.ErrNotFound before 500ing

File: handler/auth.go [hunk 5]
```diff
@@ -206,16 +196,7 @@ func (h *Handler) VerifyKey(w http.ResponseWriter, r *http.Request) {
...
-		registrationKey.User.Verified = true
-		if err := tx.Save(&registrationKey.User).Error; err != nil {
-			return err
-		}
-		if err := tx.Delete(&registrationKey).Error; err != nil {
-			return err
-		}
-		return nil
-	}); err != nil {
+	if err := h.usersRepo.Activate(registrationKey); err != nil {
```
Comment (Sigafoos) [inline handler/auth.go:204]: I think you misunderstood. I was saying to call `Activate` with the GUID itself, not the model.

File: handler/handler.go (hunk 2 above)
Comment (mattgen88) [reply handler/handler.go:26]: no, I just wanted to convey it was a repository.

File: infrastructure/subjects/subjects.go [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+func New(db *gorm.DB) *Repository {
+	return &Repository{db: db}
+}
+
+type Repository struct {
+	db *gorm.DB
+}
+
+func (r *Repository) GetByID(id string) ([]*model.Subject, error) {
```
Comment (Sigafoos) [inline infrastructure/subjects/subjects.go:22]: Do we even really care about subjects?

File: main.go [hunk 7]
```diff
@@ -10,54 +10,20 @@ import (
...
-
-	db.Model(&model.OutboxEntry{}).AddForeignKey("user_id", "users(id)", "CASCADE", "CASCADE")
-
-	db.Model(&model.Read{}).AddForeignKey("book_id", "books(open_library_id)", "CASCADE", "CASCADE")
-	db.Model(&model.Read{}).AddForeignKey("user_id", "users(id)", "CASCADE", "CASCADE")
-
-	db.Model(&model.Review{}).AddForeignKey("user_id", "users(id)", "CASCADE", "CASCADE")
-	db.Model(&model.Review{}).AddForeignKey("book_id", "books(open_library_id)", "CASCADE", "CASCADE")
-	db.Model(&model.RegistrationKey{}).AddForeignKey("user_id", "users(id)", "CASCADE", "CASCADE")
+	infrastructure.Migrate(db)
```
Comment (Sigafoos) [inline main.go:26]: I like this bit quite a lot!

File: handler/auth.go (hunk 5 above)
Comment (mattgen88) [reply handler/auth.go:204]: I most definitely did misunderstand

File: infrastructure/subjects/subjects.go (hunk 6 above)
Comment (mattgen88) [reply infrastructure/subjects/subjects.go:22]: It's not really used at the moment, but books have a list of subjects and it's a handy bit of information to do some stuff in the future around suggestions

**Pull request merged**